Fix collection drilling

davelopez opened the pull request:
Fixes #16785

This should fix the selenium test failure:

```
FAILED lib/galaxy_test/selenium/test_history_multi_view.py::TestHistoryMultiView::test_list_list_display - selenium.common.exceptions.TimeoutException: Message: Failed waiting on history item 3 to become visible, visible datasets include [#dataset-f06d708dacad69d8,#dataset-da92407cb46a61f3]. Timeout waiting on CSS selector [.multi-history-panel .content-item[data-hid="3"]] to become visible.
```
The important stuff is in the function `onViewSubCollection` inside `CollectionPanel.vue`. I'm not 100% sure this is how we are supposed to query the elements of a sub-collection i.e. by calling `/api/dataset_collections/{hdca_id}/contents/{collection_id}` where `hdca_id` is the top HDCA in the history and `collection_id` is the `DatasetCollection` ID in DCObject, but it works on my tests... it is still a little fuzzy in my head how sub-collections work :sweat_smile: 

It also fixes the store by using a composed key `[hdca_id + collection_id]` to store and retrieve elements from a collection, since just using the collection ID can create collisions between HDCAs (top level) and sub-collections (DCOs).


## How to test the changes?
- [x] This is a refactoring of components with existing test coverage.

## License
- [x] I agree to license these and all my past contributions to the core galaxy codebase under the [MIT license](https://opensource.org/licenses/MIT).

davelopez committed:
Fix sub-collection drilling

davelopez committed:
Use composed key to store/retrieve collection elements

Using just the ID can create conflicts when storing sub-collection elements.

davelopez committed:
Adapt unit tests to use new collection key

mvdbeek commented on client/src/stores/collectionElementsStore.ts:
```diff
@@ -9,24 +9,32 @@ export const useCollectionElementsStore = defineStore("collectionElementsStore",
+     * Returns a key that can be used to store or retrieve the elements of a collection in the store.
+     */
+    function getCollectionKey(collection: HDCASummary) {
+        return `${collection.id}-${collection.collection_id}`;
```
So this seems pretty weird to me. Are you storing multiple types of collections (HDCA / DatasetCollection / DatasetCollectionElement) in the store ?

These are all distinct, and in practice maybe this works, but if you're using this function on a DatasetCollectionElement you instead of a HDCA you may still get the wrong item.

davelopez replied:
Yes, I agree, and most likely I am still mixing concepts :disappointed: 
My aim was to be able to call `/api/dataset_collections/{hdca_id}/contents/{collection_id}` from any `HDCA` or `DatasetCollection` (`DatasetCollectionElements` are skipped when using `onViewSubCollection`, only their name is used along with the `object`). But I'm not even sure this is how it is supposed to query the elements of subcollections.

davelopez replied:
In the latest changes, I think I've addressed this by consistently using the `DatasetCollection` ID as the key for the elements regardless of the source collection we are querying (top-level HDCA or sub-collection)

mvdbeek commented on client/src/stores/services/index.ts:
```diff
@@ -1,9 +1,28 @@
 import { components } from "@/schema";
 
+/** Minimal representation of a collection that can contain datasets or other collections.
```
That's the definition of a DatasetCollectionElement

davelopez replied:
I have refactored a bit these types and docs and maybe it is a bit more easy to understand now.

davelopez committed:
Use collection_id as key for storing/retrieving DCEs

davelopez committed:
Refactor collection related type definitions + docs

davelopez committed:
Convert DCObjects to SubCollections when drilling

This is just to make the handling of top level HDCAs and sub-collections more consistent for fetching and storing elements in the store.

davelopez committed:
Consistently use DatasetCollection.id in the store

For storing and retrieving collection elements we only need to use the DatasetCollection ID whether the collection entry is an HDCA or a sub-collection.

davelopez committed:
Consistently fetch collection elements

Whether we are fetching elements for a top level HDCA or a sub-collection we need to provide the top HDCA ID and the DatasetCollection ID to fetch their elements.

mvdbeek commented on client/src/components/History/CurrentCollection/CollectionPanel.vue:
```diff
@@ -40,7 +40,7 @@ const dsc = computed(() => props.selectedCollections[props.selectedCollections.l
 const loading = computed(() => collectionElementsStore.isLoadingCollectionElements(dsc.value));
 const jobState = computed(() => dsc.value?.job_state_summary);
-const rootCollection = computed(() => props.selectedCollections[0]);
+const rootCollection = computed(() => props.selectedCollections[0] as HDCASummary);
```
Is that always the HDCA ? Can we do something silly like making sure there's a `history_id` attribute there and throw an error if it is not?

davelopez replied:
Done! It now checks that the root must be an HDCA otherwise it raises an error.

mvdbeek commented on client/src/components/History/CurrentCollection/CollectionPanel.vue:
```diff
@@ -59,9 +59,16 @@ function onScroll(newOffset: number) {
-    const collection = await collectionElementsStore.getCollection(itemObject.id);
-    emit("view-collection", collection);
+async function onViewSubCollection(itemObject: DCObject, name: string) {
+    // We need to convert the DCO to a SubCollection by providing
```
Would it make more sense to call this onViewDatasetCollectionElement ? That lines it up with the internal state and the objects coming out of the API. That would also mean you would pass around individual DatasetCollectionElements, which would avoid having to pass around `name`.

davelopez replied:
Done! now we pass directly the DatasetCollectionElement instead of the internal DCObject.

mvdbeek commented on client/src/stores/collectionElementsStore.ts:
```diff
@@ -1,32 +1,45 @@
+     *
+     * It consistently returns a DatasetCollection ID for (top level) HDCAs or sub-collections.
+     */
+    function getCollectionKey(collection: CollectionEntry): string {
```
We should always know what we're dealing with so we don't need to make runtime guesses here.

mvdbeek replied:
Although it is pretty explicit ... it'll work for now.

mvdbeek reviewed (approved):
Thanks for fixing this! I'd love if this was closer to the API and we didn't have to invent additional models like SubCollection, but maybe that would require having a better way to access elements of a collection.

davelopez committed:
Refactor from code review

- Make sure the computed properties are of the right type or error out.
- Rename onViewSubCollection to onViewDatasetCollectionElement.
